File: src/main/ui/src/factories/FilterOperatorGroupFactory.ts
import {FilterOperatorET} from "@/models/runtime/filter/FilterOperatorET";
import {DataTypeET} from "@/models/core/DataTypeET";

/**
 * Factory to create groups by DataTypeET to determine allowed operations, represented by FilterOperatorET values.
 * @author Kevin Link
 */
export class FilterOperatorGroupFactory {

    public createOperatorGroup(dataType: DataTypeET): FilterOperatorET[] {
        switch (dataType) {
            case DataTypeET.STRING:
                return this.createStringOperatorGroup();
            case DataTypeET.NUMBER:
                return this.createNumberOperatorGroup();
            case DataTypeET.BOOLEAN:
                return this.createBooleanOperatorGroup();
            case DataTypeET.DATE:
                return this.createDateOperatorGroup();
            case DataTypeET.ENUM:
                return this.createEnumOperatorGroup();
            default:
                throw new Error(`Unsupported data type '${dataType}' given.`);
        }
    }

    public createStringOperatorGroup(): FilterOperatorET[] {
        return [
            FilterOperatorET.EQUAL,
            FilterOperatorET.NOT_EQUAL,
            FilterOperatorET.LIKE,
            FilterOperatorET.NOT_LIKE
        ];
    }

    public createNumberOperatorGroup(): FilterOperatorET[] {
        return [
            FilterOperatorET.EQUAL,
            FilterOperatorET.NOT_EQUAL,
            FilterOperatorET.GREATER_THAN,
            FilterOperatorET.GREATER_OR_EQUAL,
            FilterOperatorET.LESS_THAN,
            FilterOperatorET.LESS_OR_EQUAL
        ];
    }

    public createBooleanOperatorGroup(): FilterOperatorET[] {
        return [
            FilterOperatorET.IS_TRUE,
            FilterOperatorET.IS_FALSE
        ];
    }

    public createDateOperatorGroup(): FilterOperatorET[] {
        return [
            FilterOperatorET.EQUAL,
            FilterOperatorET.NOT_EQUAL,
            FilterOperatorET.GREATER_THAN,
            FilterOperatorET.GREATER_OR_EQUAL,
            FilterOperatorET.LESS_THAN,
            FilterOperatorET.LESS_OR_EQUAL
        ];
    }

    public createEnumOperatorGroup(): FilterOperatorET[] {
        return [
            FilterOperatorET.EQUAL,
            FilterOperatorET.NOT_EQUAL
        ];
    }
}

File: src/main/java/de/fhswf/genericapplication/filter/data/FilterOperatorGroupFactory.java
package de.fhswf.genericapplication.filter.data;

import de.fhswf.genericapplication.dto.requests.FilterOperatorET;

/**
 * Factory to create groups of {@link FilterOperatorET} to determine allowed operators by {@link DataTypeET}.
 *
 * @author Kevin Link
 */
public class FilterOperatorGroupFactory {
    public static FilterOperatorET[] createOperatorGroup(DataTypeET dataType) {
        return switch (dataType) {
            case STRING -> createStringOperatorGroup();
            case NUMBER -> createNumberOperatorGroup();
            case BOOLEAN -> createBooleanOperatorGroup();
            case DATE -> createDateOperatorGroup();
            case ENUM -> createEnumOperatorGroup();
            case COLLECTION -> createCollectionOperatorGroup();
            case ENTITY -> createEntityOperatorGroup();
            default -> throw new RuntimeException("Unsupported data type " + dataType + " given.");
        };
    }

    public static FilterOperatorET[] createStringOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.NOT_EQUAL,
                FilterOperatorET.LIKE,
                FilterOperatorET.NOT_LIKE
        };
    }

    public static FilterOperatorET[] createNumberOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.NOT_EQUAL,
                FilterOperatorET.GREATER_THAN,
                FilterOperatorET.GREATER_OR_EQUAL,
                FilterOperatorET.LESS_THAN,
                FilterOperatorET.LESS_OR_EQUAL,
                FilterOperatorET.IN
        };
    }

    public static FilterOperatorET[] createBooleanOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.IS_TRUE,
                FilterOperatorET.IS_FALSE
        };
    }

    public static FilterOperatorET[] createDateOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.NOT_EQUAL,
                FilterOperatorET.GREATER_THAN,
                FilterOperatorET.GREATER_OR_EQUAL,
                FilterOperatorET.LESS_THAN,
                FilterOperatorET.LESS_OR_EQUAL
        };
    }

    public static FilterOperatorET[] createEnumOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.NOT_EQUAL
        };
    }

    private static FilterOperatorET[] createCollectionOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.IN
        };
    }

    private static FilterOperatorET[] createEntityOperatorGroup() {
        return new FilterOperatorET[]{
                FilterOperatorET.EQUAL,
                FilterOperatorET.NOT_EQUAL,
                FilterOperatorET.IN
        };
    }
}